lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -554,92 +642,567 @@ void crypto_bignum_clear(struct bignum *s)
+		md_algo = tee_algo_to_mbedtls_hash_algo(algo);
+		if (md_algo == MBEDTLS_MD_NONE) {
+			res = TEE_ERROR_NOT_SUPPORTED;
+			goto out;
+		}
+	}
+
+	mbedtls_rsa_set_padding(&rsa, lmd_padding, md_algo);
+
+	lmd_res = pk_info->encrypt_func(&rsa, src, src_len, dst, dst_len,
+					*dst_len, mbd_rand, NULL);
+	if (lmd_res != 0) {
+		FMSG("encrypt_func() returned 0x%x", -lmd_res);
+		res = get_tee_result(lmd_res);
+		goto out;
+	}
+	res = TEE_SUCCESS;
+out:
+	/* Reset mpi to skip freeing here, those mpis will be freed with key */
+	mbedtls_mpi_init(&rsa.E);
+	mbedtls_mpi_init(&rsa.N);
+	mbedtls_rsa_free(&rsa);
+	return res;
 }
 
-TEE_Result crypto_acipher_rsassa_sign(uint32_t algo __unused,
-				      struct rsa_keypair *key __unused,
-				      int salt_len __unused,
-				      const uint8_t *msg __unused,
-				      size_t msg_len __unused,
-				      uint8_t *sig __unused,
-				      size_t *sig_len __unused)
+TEE_Result crypto_acipher_rsassa_sign(uint32_t algo, struct rsa_keypair *key,
+				      int salt_len __unused, const uint8_t *msg,
+				      size_t msg_len, uint8_t *sig,
+				      size_t *sig_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res;
+	int lmd_res, lmd_padding;
+	size_t mod_size, hash_size;
+	mbedtls_rsa_context rsa;
+	const mbedtls_pk_info_t *pk_info = NULL;
+	uint32_t md_algo;
+
+	rsa_init_from_key_pair(&rsa, key);
+
+	switch (algo) {
+	case TEE_ALG_RSASSA_PKCS1_V1_5_MD5:
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA1:
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA224:
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA256:
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA384:
+	case TEE_ALG_RSASSA_PKCS1_V1_5_SHA512:
+		lmd_padding = MBEDTLS_RSA_PKCS_V15;
+		break;
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA1:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA224:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA384:
+	case TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA512:
+		lmd_padding = MBEDTLS_RSA_PKCS_V21;
+		break;
+	default:
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto err;
+	}
+
+	res = tee_hash_get_digest_size(TEE_DIGEST_HASH_TO_ALGO(algo),
+				       &hash_size);
+	if (res != TEE_SUCCESS)
+		goto err;
+
+	if (msg_len != hash_size) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto err;
+	}
+
+	mod_size = crypto_bignum_num_bytes(key->n);
+	if (*sig_len < mod_size) {
+		*sig_len = mod_size;
+		res = TEE_ERROR_SHORT_BUFFER;
+		goto err;
+	}
+	rsa.len = mod_size;
+
+	md_algo = tee_algo_to_mbedtls_hash_algo(algo);
+	if (md_algo == MBEDTLS_MD_NONE) {
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto err;
+	}
+
+	pk_info = mbedtls_pk_info_from_type(MBEDTLS_PK_RSA);
+	if (pk_info == NULL) {
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto err;
+	}
+
+	mbedtls_rsa_set_padding(&rsa, lmd_padding, md_algo);
+
+	if (lmd_padding == MBEDTLS_RSA_PKCS_V15)
+		lmd_res = pk_info->sign_func(&rsa, md_algo, msg, msg_len, sig,
+					     sig_len, NULL, NULL);
+	else
+		lmd_res = pk_info->sign_func(&rsa, md_algo, msg, msg_len, sig,
+					     sig_len, mbd_rand, NULL);
```

Comment:
Would it be an error to always supply `mbd_rand()`?
If not then we could simplify here.

Reply:
Same as above.

Reply:
OK